Fix/4196 speed up search bundles

AndrewJGaut opened the pull request:
### Reasons for making this change

The search_bundles function was slow primarily because it relied on nested subqueries. We updated the function to use inner joins instead, which should be much faster.

### Related issues

Fixes #4196. Also addresses #3787 (see example in comment below for that function).

Commit:
Remove unused column (is_superuser) from the User table.

Commit:
Speed up cl search by converting subqueries to inner joins

Commit:
Merge branch 'fix/3523-remove-is_superuser-from-user-table' into fix/4196-speed-up-search-bundles

Commit:
Merge branch 'master' of github.com:codalab/codalab-worksheets into fix/4196-speed-up-search-bundles

percyliang commented:
Do you have any before/after timings that you can share?

Commit:
Fix bug with certain keywords

Commit:
added in aliasing

Commit:
Run pre-commit

AndrewJGaut commented:
Sure thing!

Examples of speedup (run on prod):

Command: cl search .shared data_size=.sum

Original query:
SELECT sum(q1.num) AS sum_1 FROM (SELECT DISTINCT bundle.uuid AS uuid, bundle_metadata_1.metadata_value * 1 AS num FROM bundle_metadata AS bundle_metadata_1, bundle INNER JOIN group_bundle_permission ON bundle.uuid = group_bundle_permission.object_uuid INNER JOIN user_group ON group_bundle_permission.group_uuid = user_group.group_uuid WHERE user_group.user_id='0' AND group_bundle_permission.permission >= 1 AND bundle.uuid = bundle_metadata_1.bundle_uuid AND bundle_metadata_1.metadata_key='data_size') AS q1;
**7.52 seconds**

Updated query:
SELECT sum(q1.num) AS sum_1 FROM (SELECT DISTINCT bundle.uuid AS uuid, bundle_metadata_1.metadata_value * 1 AS num FROM bundle_metadata AS bundle_metadata_1, bundle INNER JOIN group_bundle_permission ON bundle.uuid = group_bundle_permission.object_uuid INNER JOIN user_group ON group_bundle_permission.group_uuid = user_group.group_uuid WHERE user_group.user_id=’0’ AND group_bundle_permission.permission >= 1 AND bundle.uuid = bundle_metadata_1.bundle_uuid AND bundle_metadata_1.metadata_key='data_size') AS q1;
**0.0 seconds**


Command: cl search .after=2021-08-20 .count

Original query:
SELECT count(q2.uuid) AS tbl_row_count FROM (SELECT DISTINCT bundle.uuid AS uuid FROM bundle WHERE bundle.uuid IN (SELECT bundle_metadata.bundle_uuid FROM bundle_metadata WHERE bundle_metadata.metadata_key = ‘created’ AND bundle_metadata.metadata_value >= 1629417600) LIMIT 0, 18446744073709551615) AS q2;
**27.74 seconds.**

Updated query:
SELECT count(q1.uuid) AS tbl_row_count FROM (SELECT DISTINCT bundle.uuid AS uuid FROM bundle INNER JOIN bundle_metadata AS bundle_metadata_1 ON bundle.uuid = bundle_metadata_1.bundle_uuid WHERE bundle_metadata_1.metadata_key = ‘created’ AND bundle_metadata_1.metadata_value >= 1629417600 LIMIT 0, 18446744073709551615) AS q1
**1.95 sec**

Commit:
Remove code used for testing speedups

Commit:
fix issue with mypy typing

Commit:
Remove docker_config files

Commit:
fix minor flake8 formatting issues

Commit:
Fix another formatting issue

percyliang reviewed (commented):
At a high-level, logic looks good, and it's great that we get such dramatic speedups!  I'd just be careful about the edge cases, and we should probably add some test cases for this.  Consider also splitting up the function into more modular pieces (something we've wanted to do for a while) if you think it makes sense.

Commit:
Use just one list of tuples for inner joins, add type annotations, and clean up variable names

Commit:
Fix the disjunct subquery for non-root user and the subquery for .floating to be joins as well. Add tests for both cases.

Commit:
minor formatting fixes

Commit:
Remove files added by pre-commit

Commit:
Remove package files added by pre-commit

Commit:
fix test that doesn't work in protected mode

Commit:
minor formatting change

Commit:
add type annotations

AndrewJGaut commented:
Thanks so much for the thorough review! I added more test cases to address the edge cases you mentioned (see my comment about INNER JOIN and LEFT OUTER JOIN above).

In terms of splitting the function, I do think that makes sense and gave it some thought throughout the day. However, I'm still not sure how to do it. I'll keep thinking, but in the meantime I addressed your other comments. Thanks!

epicfaace commented:
> **0.0 seconds**

Amazing!!

Commit:
revert changes made in other PR to remove the superuser column

AndrewJGaut committed:
Merge branch 'master' into fix/4196-speed-up-search-bundles

Commit:
add some comments to make the function more understandable

Commit:
Merge branch 'fix/4196-speed-up-search-bundles' of github.com:codalab/codalab-worksheets into fix/4196-speed-up-search-bundles

percyliang reviewed (approved):
Looks great!

AndrewJGaut committed:
Merge branch 'master' into fix/4196-speed-up-search-bundles

AndrewJGaut committed:
Merge branch 'master' into fix/4196-speed-up-search-bundles
